Display ingame notification if custom ruleset issues are encountered during startup.

orendain opened the pull request:
Resolves https://github.com/orendain/DemeoMods/issues/192.

- Added option to `ConfigManager.ImportRuleset()`, `bool tolerateFailures` that specifies whether or not rule-parsing failures should be tolerated.  This is currently unused, and is always set to `false`.
  - _**This means that rulesets are not partially imported. Either all configured rules are validated, or the ruleset fails.**_
- Applied the `Try-` pattern when importing rulesets, to propagate information about whether or not there were issues loading particular rulesets.
- The above allows us to provide a notification in-game when HouseRules encountered issues loading a ruleset.

Example in-game message:
![image](https://user-images.githubusercontent.com/1426304/154834925-5883e6a6-3f4b-41c4-8c0b-56c5aa7ba198.png)


Example logs of a failed config load:
![image](https://user-images.githubusercontent.com/1426304/154824381-26954b2b-d436-47bb-a961-d3b53fc432d4.png)

jimconner commented on HouseRules_Configuration/ConfigManager.cs:
```diff
@@ -105,22 +105,44 @@ internal string ExportRuleset(Ruleset ruleset)
+                }
+            }
+
+            return hadNoErrors;
```
Was wondering about returning a list of the failed [{file}]s here, so they could be displayed in-game.

orendain replied:
Done.  Moved implementation around so that now a list of files that failed to be loaded is captured.

jimconner commented:
I cannot get the in-game message to display. I tried two different errors in my json, a rulename that didn't exist, and also an abilityname that doesn't exist.

I get the `Skipping that ruleset:` in yellow (as in your screenshot above), but then my logs go into red ink with a NullReferenceException.

```
[11:17:46.765] [HouseRules:Configuration] [WARNING] Failed to import ruleset from file [D:\Occulus\Software\resolution-games-demeo\UserData\HouseRules\LuckyDip.json]. Skipping that ruleset: System.InvalidOperationException: Failed to read rule entry [PieceImmnityListAdjusted] of ruleset [LuckyDip]. ---> System.ArgumentException: Could not find a rule type represented by the name: PieceImmnityListAdjusted

  at HouseRules.Configuration.ConfigManager.FindRuleAndConfigType (System.String ruleName) [0x0002e] in <acab5ae4bf9c4367b7a2c04e81a564fd>:0 

  at HouseRules.Configuration.ConfigManager.ImportRuleset (System.String fileName, System.Boolean tolerateFailures) [0x00032] in <acab5ae4bf9c4367b7a2c04e81a564fd>:0 

   --- End of inner exception stack trace ---

  at HouseRules.Configuration.ConfigManager.ImportRuleset (System.String fileName, System.Boolean tolerateFailures) [0x000c1] in <acab5ae4bf9c4367b7a2c04e81a564fd>:0 

  at HouseRules.Configuration.ConfigManager.TryImportRulesets (System.Collections.Generic.List`1[HouseRules.Types.Ruleset]& rulesets) [0x00058] in <acab5ae4bf9c4367b7a2c04e81a564fd>:0 
[11:17:50.173] [576572:HouseRules:_Configuration] [ERROR] System.NullReferenceException: Object reference not set to an instance of an object

  at Boardgame.GameUI.ShowCameraMessage (System.String s, System.Single seconds) [0x00000] in <5a833664a0a84ffa89e018e6a3a0881a>:0 

  at HouseRules.Configuration.ConfigurationMod.ShowStartupIssuesMessage () [0x00043] in <acab5ae4bf9c4367b7a2c04e81a564fd>:0 

  at HouseRules.Configuration.ConfigurationMod.OnSceneWasInitialized (System.Int32 buildIndex, System.String sceneName) [0x00039] in <acab5ae4bf9c4367b7a2c04e81a564fd>:0 

  at MelonLoader.MelonHandler+<>c__DisplayClass45_0.<OnSceneWasInitialized>b__0 (MelonLoader.MelonMod x) [0x0000c] in <239dd078bb4e49c8a1ff8200d76f2e32>:0 

  at MelonLoader.MelonHandler.InvokeMelonMethod[T] (System.Collections.Generic.List`1[T]& melons, MelonLoader.MelonHandler+InvokeMelonMethodDelegate`1[T] method, System.Boolean remove_failed) [0x00032] in <239dd078bb4e49c8a1ff8200d76f2e32>:0 
```

orendain commented:
Yeah, that's due to a bad merge with the main branch.  The latest commit fixed it - could you try pulling anew?

jimconner commented:
Git tells me that there is nothign more to pull, on either main or this branch. My commit logs shows me that the 'Applied missed cleanup' as at the head. I rebuilt DLLs etc... still the same error.

orendain commented:
Oh ... in that case I'll take a look now!

orendain commented:
Ah - I only partially read the message and assumed it was something else.

>   at Boardgame.GameUI.ShowCameraMessage (System.String s, System.Single seconds) [0x00000] in <5a833664a0a84ffa89e018e6a3a0881a>:0 
Unfortunately, this is the issue I originally when I determined the `GameUI Camera` was not initialized by the time the game loads -- making notification messages impossible.

It started working on my client (perhaps a Steam version particularity) and I just started assuming I must have done something weird for it to not have worked before.

But if you're hitting the same issue as I originally was ... then I'll have to take a closer look and perhaps either delay the notification until the GameUI camera loads ... or pause notification for now.

orendain commented:
For future self:

> Was wondering about returning a list of the failed [{file}]s here, so they could be displayed in-game.

Done.  Moved implementation around so that now a list of files that failed to be loaded is captured.

However, I've also removed the bit that brings up a notification on the UI.  We may have to implement a wait mechanism (e.g., a while look) to wait until the message system is up and running - and it might be best to leave it for another PR so as to not block (not sure if I can tackle it soon).
 
So now:
- This PR will fail a ruleset entirely if one of its rules fails to parse.
- Failed rulesets will not show up on the ingame UI.
- Failed files/rulesets are captured, but nothing is currently done with them (until UI/notification is worked out).

Quick snippet to remind myself of the message that we had:
```
private static void ShowStartupIssuesMessage()
{
    var message = new StringBuilder()
        .AppendLine("Attention:")
        .AppendLine().AppendLine("HouseRules encountered issues loading at least one of your custom rulesets.")
        .AppendLine().AppendLine("Those rulesets have been left out. You may proceed without them.")
        .AppendLine().AppendLine("See logs for more information.");

    FailedRulesetFiles.ForEach(f => message.AppendLine(Path.GetFileName(f)));

    GameUI.ShowCameraMessage(message.ToString(), 15f);
}
```

jimconner reviewed (approved):
Sounds like it should be sufficient to me. Happy for this to be merged.
